packages/http/src/validator/validators/body.ts

```diff
@@ -1,82 +1,131 @@
+}
+
+function maybeValidateFormBody(
+  schema: JSONSchema,
+  content: IMediaTypeContent,
+  mediaType: string | undefined,
+  target: any,
+) {
+  if (mediaType && typeIs.is(mediaType, 'application/x-www-form-urlencoded')) {
+    const encodings = get(content, 'encodings', []);
+    const encodedUriParams = splitUriParams(target);
 
-    return validateAgainstSchema(target, schema).map(error =>
-      Object.assign({}, error, { path: [prefix, ...(error.path || [])] }),
+    return pipe(
+      validateAgainstReservedCharacters(encodedUriParams, encodings),
+      Either.map(decodeUriEntities),
+      Either.map(deserializeFormBody.bind(undefined, schema, encodings)),
     );
   }
 
-  private getContent(specs: IMediaTypeContent[], mediaType?: string): IMediaTypeContent | undefined {
-    if (!mediaType) {
-      return specs[0];
-    }
+  return Either.right(target);
+}
+
+function applyPrefix(prefix: string, diagnostics: IPrismDiagnostic[]): IPrismDiagnostic[] {
+  return diagnostics.map(d => ({ ...d, path: [prefix, ...(d.path || [])] }));
+}
+
+function validateAgainstReservedCharacters(encodedUriParams: Dictionary<string, string>, encodings: IHttpEncoding[]) {
+  return pipe(
+    encodings,
+    Array.reduce<IHttpEncoding, IPrismDiagnostic[]>([], (diagnostics, encoding) => {
+      const allowReserved = get(encoding, 'allowReserved', false);
+      const property = encoding.property;
+      const value = encodedUriParams[property];
 
-    const content = specs.find(c => c.mediaType === mediaType);
+      if (!allowReserved && typeof value === 'string' && value.match(/[\/?#\[\]@!$&'()*+,;=]/)) {
+        diagnostics.push({
+          path: [property],
+          message: 'Reserved characters used in request body',
+          severity: DiagnosticSeverity.Error,
+        });
+      }
 
-    if (!content) {
-      return specs[0];
-    }
+      return diagnostics;
+    }),
+    diagnostics => (diagnostics.length ? Either.left(diagnostics) : Either.right(encodedUriParams)),
+  );
+}
 
-    return content;
+function deserializeFormBody(
+  schema: JSONSchema,
+  encodings: IHttpEncoding[],
+  decodedUriParams: Dictionary<string, string>,
+) {
+  if (!schema.properties) {
+    return decodedUriParams;
   }
+
+  return pipe(
+    Object.keys(schema.properties),
+    Array.reduce({}, (deserialized, property) => {
+      deserialized[property] = decodedUriParams[property];
+      const encoding = encodings.find(enc => enc.property === property);
+
+      if (encoding && encoding.style) {
+        const deserializer = body.get(encoding.style);
+        if (deserializer && schema.properties) {
+          const propertySchema = schema.properties[property];
+          deserialized[property] = deserializer.deserialize(property, decodedUriParams, propertySchema as JSONSchema);
+        }
+      }
+
+      return deserialized;
+    }),
+  );
+}
+
+function getContent(specs: IMediaTypeContent[], mediaType?: string) {
```

Comment:
If so, I think we need to use a better name, such as `getContentByMediaTypeOrFirst` or even better, do not do `getOrElse` and return the `Option` as it is; let the caller decide the fallback mechanism

Reply:
I also think that `mediaType` should not be nullable; either you find a way to provide it or, you know, find another way.

Reply:
I changed function name as you suggested - it's way better now.

The method will always return sth, so it is not worth returning Option here. `mediaType` is nullable at entry point level (`validate` method). And it is ok since we can deal with such a situation (by choosing first available).